### PR TITLE
[AMDGPU] Enhance s_waitcnt insertion before barrier for gfx12

### PR DESCRIPTION
**Pull request by dstutt**
Code to determine if a waitcnt is required before a barrier instruction only
considered S_BARRIER.
gfx12 adds barrier_signal/wait so need to enhance the existing code to look for
a barrier start (which is just an S_BARRIER for earlier architectures).

### CODE REVIEW

**Review by jayfoad — APPROVED** [review]
LGTM with nits

File: llvm/lib/Target/AMDGPU/SIInstrInfo.h [hunk 1]
```diff
@@ -1386,6 +1386,14 @@ class SIInstrInfo final : public AMDGPUGenInstrInfo {
   // This is used if an operand is a 32 bit register but needs to be aligned
   // regardless.
   void enforceOperandRCAlignment(MachineInstr &MI, unsigned OpName) const;
+
+  bool isBarrierStart(uint16_t Opcode) const {
```
[the remark this answers is not in the document]
Comment (Pierre-vh) [reply llvm/lib/Target/AMDGPU/SIInstrInfo.h:1390]: We use unsigned for opcodes in general, I think this should also use unsigned to be consistent (though it doesn't really matter)

[the remark this answers is not in the document]
Comment (dstutt) [reply llvm/lib/Target/AMDGPU/SIInstrInfo.h]: Done

[the remark this answers is not in the document]
Comment (dstutt) [reply llvm/lib/Target/AMDGPU/SIInstrInfo.h]: > We use unsigned for opcodes in general, I think this should also use unsigned to be consistent (though it doesn't really matter)

I just made the function consistent with others in the same file - they all use uint16_t

[the remark this answers is not in the document]
Comment (dstutt) [reply llvm/lib/Target/AMDGPU/SIInstrInfo.h]: Mostly :)

[the remark this answers is not in the document]
Comment (dstutt) [reply llvm/lib/Target/AMDGPU/SIInstrInfo.h]: I changed it since MachineInstr::getOpcode() returns unsigned

**Pull request merged**